octoprint_printhistory/parser.py

```diff
@@ -0,0 +1,294 @@
+                # extract a string between semicolons ; =>target string<= ;
+                line = line.split(";")[1].strip()
+                if self.is_parameter.match(line):
+                    comma_split = line.split(",")
+                    param = comma_split[0]
+                    value = ",".join(comma_split[1:])
+                    try:
+                        parameters.update({param: value})
+                    except ValueError:
+                        pass
+            # comment section is ended
+            else:
+                break
+        file.seek(0)
+        return parameters
+
+    def parse_bottom(self, file):
+        # Filament length is a duplicate
+        # ; Build Summary
+        # ;   Build time: 3 hours 30 minutes
+        # ;   Filament length: 54599.6 mm (54.60 m)
+        # ;   Plastic volume: 131327.49 mm^3 (131.33 cc)
+        # ;   Plastic weight: 164.16 g (0.36 lb)
+        # ;   Material cost: 20.52
+        parameters = {}
+        for line in reverse_readline(file):
+            if line.startswith(";"):
+                line = line.split(";")[1].strip()
+                if self.is_bparameter.match(line):
+                    splitted = line.split(":")
+                    param = splitted[0]
+                    value = splitted[1].strip()
+                    parameters.update({param: value})
+            else:
+                break
+        # Rename "Filament length"
+        try:
+            value = parameters.pop("Filament length")
+            parameters["Filament length Gcode"] = value
+        except:
+            pass
+        return parameters
+
+
+# memory-efficient way to read file backwards
+# http://stackoverflow.com/questions/2301789/read-a-file-in-reverse-order-using-python
+def reverse_readline(fh, buf_size=8192):
+    """a generator that returns the lines of a file in reverse order"""
+    segment = None
+    offset = 0
+    fh.seek(0, os.SEEK_END)
+    file_size = remaining_size = fh.tell()
+    while remaining_size > 0:
+        offset = min(file_size, offset + buf_size)
+        fh.seek(file_size - offset)
+        buffer = fh.read(min(remaining_size, buf_size))
+        remaining_size -= buf_size
+        lines = buffer.split('\n')
+        # the first line of the buffer is probably not a complete line so
+        # we'll save it and append it to the last line of the next buffer
+        # we read
+        if segment is not None:
+            # if the previous chunk starts right from the beginning of line
+            # do not concact the segment to the last line of new chunk
+            # instead, yield the segment first
+            if buffer[-1] is not '\n':
+                lines[-1] += segment
+            else:
+                yield segment
+        segment = lines[0]
+        for index in range(len(lines) - 1, 0, -1):
+            if len(lines[index]):
+                yield lines[index]
+    # Don't yield None if the file was empty
```

Comment:
useless comment since it is explaining HOW and not WHY https://blog.codinghorror.com/code-tells-you-how-comments-tell-you-why/